File: Mario-AI-Interface/src/agents/HumanAgent.java
package agents;

import java.awt.event.KeyAdapter;
import java.awt.event.KeyEvent;

import engine.core.MarioAgent;
import engine.core.MarioForwardModel;
import engine.core.MarioTimer;
import engine.helper.MarioActions;

public class HumanAgent extends KeyAdapter implements MarioAgent {
    private boolean[] actions = null;
    private boolean oldControl = true;
    private boolean isPressed = false;

    public HumanAgent(boolean oldControl){
        this.oldControl = oldControl;
    }
    @Override
    public void initialize(MarioForwardModel model, MarioTimer timer) {
        actions = new boolean[MarioActions.numberOfActions()];
        isPressed = false;
    }

    @Override
    public boolean[] getActions(MarioForwardModel model, MarioTimer timer) {
        if(!isPressed){
            return actions.clone();
        }else{
            System.out.println("return true");
            return new boolean[]{true, false, true, true, true, true};
        }

    }

    @Override
    public String getAgentName() {
        return "HumanAgent";
    }

    @Override
    public void keyPressed(KeyEvent e) {
        toggleKey(e.getKeyCode(), true);
    }

    @Override
    public void keyReleased(KeyEvent e) {
        toggleKey(e.getKeyCode(), false);
    }

    private void toggleKey(int keyCode, boolean isPressed) {
        if (this.actions == null) {
            return;
        }
        if(keyCode==KeyEvent.VK_0){
            if(isPressed){
                this.isPressed = true;
                System.out.println("Pressed");
            }else{
                this.isPressed = false;
            }
        }
        if(oldControl){
            switch (keyCode) {
                case KeyEvent.VK_LEFT:
                    this.actions[MarioActions.LEFT.getValue()] = isPressed;
                    break;
                case KeyEvent.VK_RIGHT:
                    this.actions[MarioActions.RIGHT.getValue()] = isPressed;
                    break;
                case KeyEvent.VK_DOWN:
                    this.actions[MarioActions.DOWN.getValue()] = isPressed;
                    break;
                case KeyEvent.VK_S:
                    this.actions[MarioActions.JUMP.getValue()] = isPressed;
                    break;
                case KeyEvent.VK_A:
                    this.actions[MarioActions.SPEED.getValue()] = isPressed;
                    break;
            }
        }
        else{
            switch (keyCode) {
                case KeyEvent.VK_A:
                    this.actions[MarioActions.LEFT.getValue()] = isPressed;
                    break;
                case KeyEvent.VK_D:
                    this.actions[MarioActions.RIGHT.getValue()] = isPressed;
                    break;
                case KeyEvent.VK_S:
                    this.actions[MarioActions.DOWN.getValue()] = isPressed;
                    break;
                case KeyEvent.VK_J:
                    this.actions[MarioActions.JUMP.getValue()] = isPressed;
                    break;
                case KeyEvent.VK_K:
                    this.actions[MarioActions.SPEED.getValue()] = isPressed;
                    break;
            }
        }

    }

}

File: templates/logics/agents/HumanAgent.js
import MarioAgent from "../engine/core/MarioAgent.js";
import {MarioActions} from "../engine/helper/MarioActions.js";

const W = 'w'
const S = 's'
const A = 'a'
const D = 'd'
const J = 'j'
const K = 'k'
const Z = 'z'
const X = 'x'
const UP = 'arrowUp'
const DOWN = 'arrowDown'
const LEFT = 'arrowLeft'
const RIGHT = 'arrowRight'


const actionMap = {
    // WASD
    A: MarioActions.LEFT[0],
    D: MarioActions.RIGHT[0],
    S: MarioActions.DOWN[0],
    J: MarioActions.JUMP[0],
    K: MarioActions.SPEED[0],

    //ARROW
    DOWN: MarioActions.DOWN[0],
    LEFT: MarioActions.LEFT[0],
    RIGHT: MarioActions.RIGHT[0],
    X: MarioActions.SPEED[0],
    Z: MarioActions.JUMP[0],
}

export default class HumanAgent extends MarioAgent {
    actions;
    isPressed = false;
    p = 0;

    gameScene ;

    setGameScene(){

    }

    constructor(gameScene) {
        super();
        this.actions = new Array(5)
        // gameScene.input.keyboard.on('keydown', event => {
        //     this.toggleKey(event.key, true)
        // });

        // gameScene.input.keyboard.on('keyup', event => {
        //     this.toggleKey(event.key, false)
        // });
    }

    toggleKey(key, isPressed) {
        switch (key) {
            case A:
            case LEFT:
                this.actions[MarioActions.LEFT[0]] = isPressed
                break;
            case D:
            case RIGHT:
                this.actions[MarioActions.RIGHT[0]] = isPressed
                break;
            case S:
            case DOWN:
                this.actions[MarioActions.DOWN[0]] = isPressed
                break;
            case K:
            case X:
                this.actions[MarioActions.SPEED[0]] = isPressed
                break;
            case J:
            case Z:
                this.actions[MarioActions.JUMP[0]] = isPressed
                break;
        }
    }

    getActions(model, timer) {
        if (!this.isPressed) {
            return JSON.parse(JSON.stringify(this.actions))
        } else {
            return [true, false, true, true, true, true]
        }
    }

    getAgentName() {
        return "HumanAgent"
    }

}